File: client/src/domain/profit.ts
// Mirror of the server's profit math (ResellTracker.Domain/Profit.cs), used for
// the sale form's live breakdown. The server stays the source of truth; both
// implementations run shared/profit-cases.json so they cannot drift.

import { add, cents, divideRoundHalfUp, fromDollars, sub, ZERO, type Cents } from './money'
import { FALLBACK_PLATFORM, scheduleFor, type FeeSettings } from './platforms'

/** What a sale recorded. `payout` null means "not known yet". */
export interface SaleFigures {
  platform: string
  price: Cents
  shippingCharged: Cents
  payout: Cents | null
  shippingCost: Cents
  otherCosts: Cents
}

export interface ProfitBreakdown {
  gross: Cents
  payout: Cents
  fees: Cents
  cogs: Cents
  shippingCost: Cents
  otherCosts: Cents
  costs: Cents
  net: Cents
  /** Net over gross; null with no revenue. */
  margin: number | null
  /** Net over cost of goods; null for an item that cost nothing. */
  roi: number | null
  feesEstimated: boolean
}

export interface ProfitTotals {
  gross: Cents
  payout: Cents
  fees: Cents
  costs: Cents
  net: Cents
  count: number
  estimated: number
}

/**
 * What the platform's rate table says a sale should cost. Only ever a stand-in
 * for the real figure: see computeProfit.
 */
export function estimateFees(
  sale: { platform: string | null; price: Cents; shippingCharged: Cents },
  fees: FeeSettings,
): Cents {
  const schedule = scheduleFor(fees, sale.platform)
  const base = schedule.includesShipping ? sale.price + sale.shippingCharged : sale.price
  if (base <= 0) return ZERO

  // Percent has at most 2 decimals, so it is a whole number of hundredths of a
  // percent, and base * that / 10000 is the fee in cents before rounding.
  const hundredthsOfPercent = Math.round(schedule.percent * 100)
  const percentPart = divideRoundHalfUp(base * hundredthsOfPercent, 10000)
  return add(cents(percentPart), fromDollars(schedule.fixed))
}

/**
 * gross = what the buyer paid; net = payout less cost of goods, shipping and other
 * costs. The payout is the pivot: entered, the fee is derived from it and is exact;
 * left null, the rate table estimates the fee and the payout follows.
 */
export function computeProfit(cost: Cents, sale: SaleFigures | null, fees: FeeSettings): ProfitBreakdown {
  const gross = add(sale?.price ?? ZERO, sale?.shippingCharged ?? ZERO)

  const payoutKnown = sale?.payout !== null && sale?.payout !== undefined
  let payout: Cents
  let platformFees: Cents
  if (payoutKnown) {
    payout = sale.payout!
    platformFees = sub(gross, payout)
  } else {
    platformFees = estimateFees(
      { platform: sale?.platform ?? null, price: sale?.price ?? ZERO, shippingCharged: sale?.shippingCharged ?? ZERO },
      fees,
    )
    payout = sub(gross, platformFees)
  }

  const shippingCost = sale?.shippingCost ?? ZERO
  const otherCosts = sale?.otherCosts ?? ZERO
  const costs = add(cost, shippingCost, otherCosts)
  const net = sub(payout, costs)

  return {
    gross,
    payout,
    fees: platformFees,
    cogs: cost,
    shippingCost,
    otherCosts,
    costs,
    net,
    margin: gross > 0 ? net / gross : null,
    roi: cost > 0 ? net / cost : null,
    feesEstimated: !payoutKnown,
  }
}

export function totalProfit(sales: readonly ProfitBreakdown[]): ProfitTotals {
  return sales.reduce<ProfitTotals>(
    (t, p) => ({
      gross: add(t.gross, p.gross),
      payout: add(t.payout, p.payout),
      fees: add(t.fees, p.fees),
      costs: add(t.costs, p.costs),
      net: add(t.net, p.net),
      count: t.count + 1,
      estimated: t.estimated + (p.feesEstimated ? 1 : 0),
    }),
    { gross: ZERO, payout: ZERO, fees: ZERO, costs: ZERO, net: ZERO, count: 0, estimated: 0 },
  )
}

/**
 * Donated stock never earns its cost back, so that cost is a write-off, kept
 * apart from sale profit rather than folded into it.
 */
export function writeOffTotal(donatedCosts: readonly Cents[]): { cost: Cents; count: number } {
  return { cost: add(...donatedCosts), count: donatedCosts.length }
}

/**
 * What a listing would net at its asking price on the first platform it is listed
 * on. Shipping is unknown before it sells, so this is price only.
 */
export function projectedNet(
  item: { listPrice: Cents | null; cost: Cents; platforms: readonly string[] },
  fees: FeeSettings,
): Cents | null {
  const price = item.listPrice
  if (price === null || price <= 0) return null
  const platform = item.platforms[0] ?? FALLBACK_PLATFORM
  const platformFees = estimateFees({ platform, price, shippingCharged: ZERO }, fees)
  return sub(sub(price, platformFees), item.cost)
}

/** How far under the asking price the accepted offer landed, or null if it didn't. */
export function markdown(listedFor: Cents, accepted: Cents): { amount: Cents; fraction: number } | null {
  if (listedFor <= 0 || accepted <= 0 || accepted >= listedFor) return null
  return { amount: sub(listedFor, accepted), fraction: (listedFor - accepted) / listedFor }
}

File: server/src/ResellTracker.Domain/Profit.cs
namespace ResellTracker.Domain;

/// <summary>What a sale recorded. <see cref="Payout"/> null means "not known yet".</summary>
public sealed record SaleFigures(
    string Platform,
    decimal Price,
    decimal ShippingCharged,
    decimal? Payout,
    decimal ShippingCost,
    decimal OtherCosts);

/// <summary>
/// Full profit breakdown for one sold item. <see cref="Margin"/> is net over gross
/// (null with no revenue); <see cref="Roi"/> is net over cost of goods (null for
/// an item that cost nothing, since dividing by zero isn't a useful number).
/// </summary>
public sealed record ProfitBreakdown(
    decimal Gross,
    decimal Payout,
    decimal Fees,
    decimal Cogs,
    decimal ShippingCost,
    decimal OtherCosts,
    decimal Costs,
    decimal Net,
    decimal? Margin,
    decimal? Roi,
    bool FeesEstimated);

public sealed record ProfitTotals(
    decimal Gross,
    decimal Payout,
    decimal Fees,
    decimal Costs,
    decimal Net,
    int Count,
    int Estimated);

public sealed record WriteOff(decimal Cost, int Count);

/// <summary>
/// The profit math. The server is the source of truth for money; the client
/// mirrors this for live preview only, and both run shared/profit-cases.json.
/// </summary>
public static class Profit
{
    /// <summary>
    /// What the platform's rate table says a sale should cost. Only ever a stand-in
    /// for the real figure: see <see cref="Compute"/>.
    /// </summary>
    public static decimal EstimateFees(string? platform, decimal price, decimal shippingCharged, FeeSettings fees)
    {
        var schedule = fees.For(platform);
        var feeBase = schedule.IncludesShipping ? price + shippingCharged : price;
        if (feeBase <= 0)
        {
            return 0;
        }

        return Money.RoundToCent(feeBase * schedule.Percent / 100m) + schedule.Fixed;
    }

    /// <summary>
    /// gross = what the buyer paid; net = payout less cost of goods, shipping and
    /// other costs. The payout is the pivot: entered, the fee is derived from it and
    /// is exact; left null, the rate table estimates the fee and the payout follows.
    /// </summary>
    public static ProfitBreakdown Compute(decimal cost, SaleFigures? sale, FeeSettings fees)
    {
        var price = sale?.Price ?? 0;
        var shippingCharged = sale?.ShippingCharged ?? 0;
        var gross = price + shippingCharged;

        decimal payout;
        decimal platformFees;
        var payoutKnown = sale?.Payout is not null;
        if (payoutKnown)
        {
            payout = sale!.Payout!.Value;
            platformFees = gross - payout;
        }
        else
        {
            platformFees = EstimateFees(sale?.Platform, price, shippingCharged, fees);
            payout = gross - platformFees;
        }

        var shippingCost = sale?.ShippingCost ?? 0;
        var otherCosts = sale?.OtherCosts ?? 0;
        var costs = cost + shippingCost + otherCosts;
        var net = payout - costs;

        return new ProfitBreakdown(
            Gross: gross,
            Payout: payout,
            Fees: platformFees,
            Cogs: cost,
            ShippingCost: shippingCost,
            OtherCosts: otherCosts,
            Costs: costs,
            Net: net,
            Margin: gross > 0 ? net / gross : null,
            Roi: cost > 0 ? net / cost : null,
            FeesEstimated: !payoutKnown);
    }

    public static ProfitTotals Total(IEnumerable<ProfitBreakdown> sales)
    {
        var totals = new ProfitTotals(0, 0, 0, 0, 0, 0, 0);
        foreach (var p in sales)
        {
            totals = new ProfitTotals(
                totals.Gross + p.Gross,
                totals.Payout + p.Payout,
                totals.Fees + p.Fees,
                totals.Costs + p.Costs,
                totals.Net + p.Net,
                totals.Count + 1,
                totals.Estimated + (p.FeesEstimated ? 1 : 0));
        }

        return totals;
    }

    /// <summary>
    /// Donated stock never earns its cost back, so that cost is a write-off. Kept
    /// apart from sale profit: "I made $40 on that jacket" and "I gave up on $18 of
    /// stock" are two different facts.
    /// </summary>
    public static WriteOff WriteOffTotal(IEnumerable<decimal> donatedCosts)
    {
        var list = donatedCosts.ToList();
        return new WriteOff(list.Sum(), list.Count);
    }

    /// <summary>
    /// What a listing would net at its asking price on the first platform it is
    /// listed on. Shipping is unknown before it sells, so this is price only.
    /// </summary>
    public static decimal? ProjectedNet(decimal? listPrice, decimal cost, IReadOnlyList<string> platforms, FeeSettings fees)
    {
        if (listPrice is not { } price || price <= 0)
        {
            return null;
        }

        var platform = platforms.Count > 0 ? platforms[0] : Platforms.Fallback.Id;
        return price - EstimateFees(platform, price, 0, fees) - cost;
    }
}
